### DISCUSSION TITLE
Open Dynamic Links from desktop

### DISCUSSION THREAD

**Discussion by whalemare** [invertase/react-native-firebase#4479] [Ideas] [answered]
### Use case

As user, I want to share my `post` in app with friends. I create dynamic link and share it. 
When my friends open this link from mobile, if app not installed, user will be redirected to store, otherwise user can see post in already installed.

When my friend try to open the same link from desktop, he will see firebase blank page with description about invalid deeplink.
```
Invalid Dynamic Link
Requested URL (https://myapp.page.link/?someDeepLink=params) must be a parsable and complete DynamicLink.

If you are the developer of this app, ensure that your Dynamic Links domain is correctly configured and that the path component of this URL is valid.
```

### Expect

As developer, I expected that when user open this link from desktop (or other platform), he will see some fallback url (my website, market link or something else), describes that this link available only from supported platforms.

### Research

In [Firebase Dynamic Links docs](https://firebase.google.com/docs/dynamic-links/create-manually) we can see, that this behavior is supported by passing specific link to `ofl` parameter, but this feature is not available on native sdk, **but** it`s available for manually creating [like this](https://stackoverflow.com/a/58874375/5662244)

For approve this, we can create [sample of long dynamic link](https://myholiday.page.link/?isi=1507764516&ifl=https%3A%2F%2Fgoogle%2Ecom&afl=https%3A%2F%2Fgoogle%2Ecom&ibi=ru%2Ewhalemare%2Emyholidays&link=https%3A%2F%2Fmyholiday%2Epage%2Elink%3FeventPath%3Dpackets%252Fworld%252Fevents%252FfT6NLYEqrbno2YiI9sxY%26linkAuthorId%3DKA3yaLjyq7YUktv3NbpTCxQh25W2%26appVersion%3D1%2E6%2E8%2E83%26title%3D%25D0%25A5%25D1%258D%25D0%25BB%25D0%25BB%25D0%25BE%25D1%2583%25D0%25B8%25D0%25BD%26description%3D31%2520%25D0%25BE%25D0%25BA%25D1%2582%25D1%258F%25D0%25B1%25D1%2580%25D1%258F%250A%25D0%25A1%25D0%25BE%25D0%25B2%25D1%2580%25D0%25B5%25D0%25BC%25D0%25B5%25D0%25BD%25D0%25BD%25D1%258B%25D0%25B9%2520%25D0%25BC%25D0%25B5%25D0%25B6%25D0%25B4%25D1%2583%25D0%25BD%25D0%25B0%25D1%2580%25D0%25BE%25D0%25B4%25D0%25BD%25D1%258B%25D0%25B9%2520%25D0%25BF%25D1%2580%25D0%25B0%25D0%25B7%25D0%25B4%25D0%25BD%25D0%25B8%25D0%25BA%252C%2520%25D0%25B2%25D0%25BE%25D1%2581%25D1%2585%25D0%25BE%25D0%25B4%25D1%258F%25D1%2589%25D0%25B8%25D0%25B9%2520%25D0%25BA%2520%25D1%2582%25D1%2580%25D0%25B0%25D0%25B4%25D0%25B8%25D1%2586%25D0%25B8%25D1%258F%25D0%25BC%2520%25D0%25B4%25D1%2580%25D0%25B5%25D0%25B2%25D0%25BD%25D0%25B8%25D1%2585%2520%25D0%25BA%25D0%25B5%25D0%25BB%25D1%258C%25D1%2582%25D0%25BE%25D0%25B2%2520%25D0%2598%25D1%2580%25D0%25BB%25D0%25B0%25D0%25BD%25D0%25B4%25D0%25B8%25D0%25B8%2520%25D0%25B8%2520%25D0%25A8%25D0%25BE%25D1%2582%25D0%25BB%25D0%25B0%25D0%25BD%25D0%25B4%25D0%25B8%25D0%25B8%2E%2520%25D0%259E%25D1%2582%25D0%25BC%25D0%25B5%25D1%2587%25D0%25B0%25D0%25B5%25D1%2582%25D1%2581%25D1%258F%2520%25D0%25BA%25D0%25B0%25D0%25B6%25D0%25B4%25D1%258B%25D0%25B9%2520%25D0%25B3%25D0%25BE%25D0%25B4%25E2%2580%25A6%26day%3D31%26month%3D10%26year%3Dundefined&si=https%3A%2F%2Ffirebasestorage%2Egoogleapis%2Ecom%2Fv0%2Fb%2Fmyholiday%2De2e4%2Eappspot%2Ecom%2Fo%2FKA3yaLjyq7YUktv3NbpTCxQh25W2%252Fevents%252F19546D37%2DB489%2D46CC%2DA0DA%2DAFBF3BF0B4C0%3Falt%3Dmedia%26token%3Dcb5708e9%2D62f1%2D4a2d%2D9398%2Dc664b3f872c4&sd=31%20%D0%BE%D0%BA%D1%82%D1%8F%D0%B1%D1%80%D1%8F%0A%D0%A1%D0%BE%D0%B2%D1%80%D0%B5%D0%BC%D0%B5%D0%BD%D0%BD%D1%8B%D0%B9%20%D0%BC%D0%B5%D0%B6%D0%B4%D1%83%D0%BD%D0%B0%D1%80%D0%BE%D0%B4%D0%BD%D1%8B%D0%B9%20%D0%BF%D1%80%D0%B0%D0%B7%D0%B4%D0%BD%D0%B8%D0%BA%2C%20%D0%B2%D0%BE%D1%81%D1%85%D0%BE%D0%B4%D1%8F%D1%89%D0%B8%D0%B9%20%D0%BA%20%D1%82%D1%80%D0%B0%D0%B4%D0%B8%D1%86%D0%B8%D1%8F%D0%BC%20%D0%B4%D1%80%D0%B5%D0%B2%D0%BD%D0%B8%D1%85%20%D0%BA%D0%B5%D0%BB%D1%8C%D1%82%D0%BE%D0%B2%20%D0%98%D1%80%D0%BB%D0%B0%D0%BD%D0%B4%D0%B8%D0%B8%20%D0%B8%20%D0%A8%D0%BE%D1%82%D0%BB%D0%B0%D0%BD%D0%B4%D0%B8%D0%B8%2E%20%D0%9E%D1%82%D0%BC%D0%B5%D1%87%D0%B0%D0%B5%D1%82%D1%81%D1%8F%20%D0%BA%D0%B0%D0%B6%D0%B4%D1%8B%D0%B9%20%D0%B3%D0%BE%D0%B4%E2%80%A6&st=%D0%A5%D1%8D%D0%BB%D0%BB%D0%BE%D1%83%D0%B8%D0%BD&apn=ru%2Ewhalemare%2Emyholidays) that correctly opens from mobile app and fallback to `google.com` from desktop.

But this link is terrible for user experience, we need short link.

### Suggestion

Since native libs do not allow to create short links with `ofl`, we can solve this problem by:
1. Generate long dynamic link and append it with ofl
2. Short it with native library api [like this](https://stackoverflow.com/a/58874375/5662244)

For now, it can be achieved with some hacky way:
1. Generate long dynamic link and append with ofl, where deeplink = that can be resolved by app
2. Generate short dynamic link, where deeplink = generated long link.
So, when user tap on this short link, he will be redirected to long dynamic link, that redirect him to our app or `ofl`.

### Question

Can we achieve the same result in more convenient way and remove hacky step?

**Comment by mikehardy** [comment, marked as answer]
My app works just fine with fallback links, here's my profile, should render correctly

https://app.komp.ai/me/29hQ

Here's the debug mode

https://app.komp.ai/me/29hQ?d=1

I generate that deep link in my app dynamically 

```typescript

      console.log('DynamicLinkStore::getProfileLink - building link');
      firebase
        .dynamicLinks()
        .buildShortLink(
          {
            link: `${AppConfig.getWebAppURIRoot()}/${user.profileName}`,
            domainUriPrefix: AppConfig.getDeepLinkRoot(),
            android: {
              packageName: AppConfig.getAndroidPackage(),
            },
            ios: {
              bundleId: AppConfig.getIOSBundle(),
              appStoreId: AppConfig.getIOSAppId(),
            },
            social: {
              title: `${user.profileName} ${I18NService.translate(
                'ProfileShareSocialYour'
              )} Komp.ai`,
              descriptionText: I18NService.translate('ProfileShareSocialDescription'),
              imageUrl: `${AppConfig.getWebAppURIRoot()}/ms-icon-310x310.png`,
            },
          },
          firebase.dynamicLinks.ShortLinkType.SHORT
        )
        .then((link) => {
          // HERE IS THE LINK HERE... 
          console.log(`DynamicLinkStore::getProfileLink - got link: ${link}`);
        })
        .catch((err) => {
          console.log('DynamicLinkStore::getProfileLink - unable to build link?', err);
          DynamicLinkStore.profileLinkLoading = false;
        });
```

...so I think everything is working

**Reply by giorgiofellipe** [reply]
What if the app doesn't have a web version at all? Mine doens't for eg., so the `ofl` parameter seems handy to these cases, I just want to redirect the user to my landing page when the link has been visited from other platforms rather than iOS and Android.

**Reply by mikehardy** [reply]
@giorgiofellipe send the user any place you like, you are in complete control of the link you put there.

**Reply by yudizBrijesh** [reply]
> What if the app doesn't have a web version at all? Mine doens't for eg., so the `ofl` parameter seems handy to these cases, I just want to redirect the user to my landing page when the link has been visited from other platforms rather than iOS and Android.

Can you share the code sippet for the same?

**Reply by mhesham32** [reply]
we have an iOS app and no Android app ofl doesn't redirect me to the specified link works only on desktop

**Comment by ritzblogs** [comment]
https://abodeofmystories.blogspot.com/2021/03/react-native-dynamic-links-from-desktop.html
